lib/shared/protocol.ts

```diff
@@ -0,0 +1,85 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** @fileoverview TODO */
+
+import { isArray } from "./types";
+
+/** TODO */
+export type FledgeRequest =
+  | [RequestTag.JOIN_AD_INTEREST_GROUP, JoinAdInterestGroupRequest]
+  | [RequestTag.LEAVE_AD_INTEREST_GROUP, LeaveAdInterestGroupRequest]
+  | [RequestTag.RUN_AD_AUCTION, RunAdAuctionRequest];
+
+/** TODO */
+export enum RequestTag {
+  JOIN_AD_INTEREST_GROUP,
+  LEAVE_AD_INTEREST_GROUP,
+  RUN_AD_AUCTION,
+}
+
+/** TODO */
+export type JoinAdInterestGroupRequest = [
+  name: string,
+  ads: Array<[renderingUrl: string, cpmInUsd: number]>
```

Comment:
Why can't this reuse `public_api.ts:Ad`?

Reply:
My understanding is that we want it to be possible for a frame that was built with tsickle to interoperate with a library that wasn't. So, for storage and for cross-origin communication, I'm avoiding object literals that require us to be careful about property renaming across multiple binaries, in favor of an array-based wire format similar to [the one used to represent proto messages inside Google](https://github.com/protocolbuffers/protobuf/issues/1591#issuecomment-237014741).

Reply:
Wouldn't the alternative just be to use `declare` to forbid property renaming on these?

Reply:
We could do that; it's a bit less compact in terms of code size, etc. (which is why the aforementioned proto wire format uses arrays), but it would work. Do you think this is preferable to the array-based wire format?

Reply:
I guess this is communication where we control both endpoints, and we check versions.  Array SGTM.